### PR TITLE
Fix 401/403 apiserver errors do not return 'Status' objects

### LINKED ISSUE #45970: 401/403 apiserver errors do not return 'Status' objects
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No.

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): "status forbidden unauthorized"

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug

**Kubernetes version** (use `kubectl version`):

Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.7", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**: minikube 0.18.0 / kubernetes 1.5.3
- **OS** (e.g. from /etc/os-release): 
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:

I'm trying to query the API server and am getting unexpected results like `Forbidden: "/apis"` when not providing (enough) credentials.

**What you expected to happen**:

According to https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#http-status-codes I expected the server to return a Status object with status = 'Failure' and code = 403/401.

**How to reproduce it** (as minimally and precisely as possible):

`curl -k API-SERVER-URL/apis`

**Anything else we need to know**:

I'm writing a client library, and want to handle errors properly, so any additional help/information or insights into the why of this behavior are very welcome as well :)

### PR DESCRIPTION
**Pull request by shiywang**
fixes https://github.com/kubernetes/kubernetes/issues/45970

### CODE REVIEW

**Comment by sttts** [comment]
@shiywang looks good. Can you add tests for both filter cases? Just add them below https://github.com/kubernetes/kubernetes/blob/ef7d63818fab7aeb3d356b46d3b0dddfd524b9c3/test/integration/master/master_test.go#L114-L114.

**Comment by shiywang** [comment]
@sttts testcase added, I also add another filter cases, but seems one unit test failed due to here https://github.com/kubernetes/kubernetes/pull/47384/commits/0025b0760177262f6239992da7cab397d74a411b#diff-6445bd63d178373b50fc083b9a043d0bR72

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 1]
```diff
@@ -57,7 +63,23 @@ func WithAuthentication(handler http.Handler, mapper genericapirequest.RequestCo
...
+					return
+				}
+
+				requestInfo, found := request.RequestInfoFrom(ctx)
+				if !found {
+					responsewriters.InternalError(w, req, errors.New("no RequestInfo found in the context"))
+					return
+				}
+
+				gv := schema.GroupVersion{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: would move this below the if

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go [hunk 2]
```diff
@@ -55,8 +59,10 @@ func WithAuthorization(handler http.Handler, requestContextMapper request.Reques
 			return
 		}
 
+		gv := schema.GroupVersion{Group: attributes.GetAPIGroup(), Version: attributes.GetAPIVersion()}
+		gr := schema.GroupResource{Group: attributes.GetAPIGroup(), Resource: attributes.GetResource()}
 		glog.V(4).Infof("Forbidden: %#v, Reason: %q", req.RequestURI, reason)
-		responsewriters.Forbidden(attributes, w, req, reason)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go]: would it be feasible to modify `responsewriters.Forbidden` with the `Status` code? There might be more places where we call it.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 3]
```diff
@@ -76,24 +98,6 @@ func WithAuthentication(handler http.Handler, mapper genericapirequest.RequestCo
 	)
 }
 
-func Unauthorized(supportsBasicAuth bool) http.HandlerFunc {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: what about adding the mapper and serializer to this function? Then we can keep the (now failing) unit tests.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go (hunk 2 above)
Comment (shiywang) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go]: hi @sttts I try to modify `responsewriters.Forbidden`, but I think that would make it almost like `responsewriters.ErrorNegotiated`, so do you think we can keep it ?  I checked, and there's no other places invoke `responsewriters.Forbidden`

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go (hunk 3 above)
Comment (shiywang) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: updated, how do you think about now ?

Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: looks good

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go (hunk 2 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go]: There is this `msg := sanitizer.Replace(forbiddenMessage(attributes))` line. We should preserve that. How do the error messages differ now from the old message?

File: test/integration/master/master_test.go [hunk 4]
```diff
@@ -101,38 +122,82 @@ func TestEmptyList(t *testing.T) {
...
+			t.Errorf("unexpected apiVersion %q, expected %q", got, expected)
+		}
+		if got, expected := decodedData["kind"], "Status"; got != expected {
+			t.Errorf("unexpected kind %q, expected %q", got, expected)
+		}
+		if got, expected := decodedData["status"], "Failure"; got != expected {
+			t.Errorf("unexpected status %q, expected %q", got, expected)
+		}
+		if got, expected := decodedData["code"], float64(tc.statusCode); got != expected {
+			t.Errorf("unexpected code %v, expected %v", got, expected)
```
Comment (sttts) [inline test/integration/master/master_test.go]: can we also test `reason` and `message`?

Comment (shiywang) [inline test/integration/master/master_test.go]: done

**Comment by shiywang** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by shiywang** [comment]
@sttts updated, ptal

**Comment by shiywang** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by shiywang** [comment]
@deads2k @sttts  could you please take a look ?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/errors_test.go [hunk 5]
```diff
@@ -64,18 +67,20 @@ func TestForbidden(t *testing.T) {
 		attributes authorizer.Attributes
 		reason     string
 	}{
-		{`User "NAME" cannot GET path "/whatever".`,
+		{"{\"metadata\":{},\"status\":\"Failure\",\"message\":\" \\\"\\\" is forbidden: User \\\"NAME\\\" cannot GET path \\\"/whatever\\\".\",\"reason\":\"Forbidden\",\"details\":{},\"code\":403}\n",
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/errors_test.go]: With  backticks you can make this more readible.

File: test/integration/master/master_test.go [hunk 6]
```diff
@@ -101,38 +122,96 @@ func TestEmptyList(t *testing.T) {
...
+			reason:       "NotFound",
+			message:      "jobs.batch \"foo\" not found",
+		},
+		{
+			name:         "403",
+			masterConfig: initStatusForbiddenMasterCongfig(),
+			statusCode:   http.StatusForbidden,
+			reqPath:      "/apis",
+			reason:       "Forbidden",
+			message:      " \"\" is forbidden: User \"\" cannot get path \"/apis\".: \"Everything is forbidden.\"",
```
Comment (sttts) [inline test/integration/master/master_test.go]: backticks

**Comment by sttts** [comment]
Looks good overall.

/cc @k8s-mirror-api-machinery-api-reviews please take a look that this is what we want.

@smarterclayton @liggitt do you see any issues with existing clients? This changes how errors are returned. Also: do we disclose anything with these more verbose 403 errors that we don't want to?

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/errors_test.go (hunk 5 above)
Comment (shiywang) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/errors_test.go]: done

File: test/integration/master/master_test.go (hunk 6 above)
Comment (shiywang) [inline test/integration/master/master_test.go]: done

**Comment by smarterclayton** [comment]
I think that because we are returning more structured errors, and there was no guarantee that another proxy couldn't be introduced (such as api aggregation), anyone expecting a generic error was already broken.  In practice, most people are either gracefully handling our structured error (but would be surprised if we returned a generic error), can handle both, or check status code first and everything else is gravy.

So I don't think this has an issue from API compat that (for instance) changing 200 -> 201 did.

**Comment by smarterclayton** [comment]
It looks like this handles negotiation properly, which was my primary concern (took a while in 1.5 to get that sorted).  So at first glance I don't see anything concerning.

**Comment by sttts** [comment]
@shiywang please squash

**Comment by shiywang** [comment]
@sttts ok to merge ?

**Pull request merged**